### ISSUE TITLE
Optional padding to hide length of ID_CRED_I and ID_CRED_R?

### ISSUE THREAD

**Issue by emanjon** [lake-wg/edhoc#189]
The confidentiality protection in EDHOC does not conceal the length of the identifiers ID_CRED_I and ID_CRED_R. We should maybe add OPTIONAL padding for deployments wanting to give eavesdroppers less information regarding the length of these identifiers. This could quite easily be achieved by allowign 0 or more simple values (e.g. true) in the beginning of the plaintexts.

```
OLD:
plaintext = ( ID_CRED_R / bstr / int, Signature_or_MAC_2, ? EAD_2 )

NEW:
plaintext = ( *true, ID_CRED_R / bstr / int, Signature_or_MAC_2, ? EAD_2 )

```

**Comment by emanjon** [comment]
I made a PR. https://github.com/lake-wg/edhoc/pull/190

The document definitly need to information on that the length of identities can be leaked and that this can be mitigated by e.g., chosing fixed length identifier for all the Initiators.

I think OPTIONAL padding should be added for use cases like certificates. I am not sure it will be used very much, but if feels like a feature a security protocol is expected to have.

**Comment by emanjon** [comment]
Based on Carsten's comment in COSE about failure from kid with int,

- My initial proposal was to make padding support for both sending and receiving optional.
- Marco commented on the PR and wanted mandatory support on the receiving side.
- Marco commented during the interim that he also wanted mandatory support on the sending side.

Before I update the PR, I think this should be discussed a bit more. There has earlier been comments from @chrysn and @StefanHri to not increase code size. I agree with Marco that mandatory to support for receiver makes sense. I am not sure about mandatory support for sender. My understanding that code for adding and removing padding would be different pieces of code.

**Comment by marco-tiloca-sics** [comment]
I proposed:

- For the receiver, mandatory to support, since a padded message might legitimately come at any time.
- For the sender, optional to use, on a per-message basis. If the sender is surely never going to pad an outgoing message, no need for it to support padding as a sender at all, of course.

**Comment by emanjon** [comment]
>For the sender, optional to use

I think the important question is if it is mandatory to implement.

**Comment by gselander** [comment]
Currently the spec says
"Padding is OPTIONAL to support."

We proposed in the interim 

”OPTIONAL to support when sending 
MANDATORY to support when received"

What is the preferred text?

**Comment by emanjon** [comment]
@marco-tiloca-sics 

The optional to use per message is already in the PR I think as "may be used".

Reading "For the sender, optional to use, on a per-message basis. If the sender is surely never going to pad an outgoing message, no need for it to support padding as a sender at all, of course." It seem like you agree with the suggestion proposed at the interim. We understood it as you proposed changes.

The "may be used" could be changes to "optional...." if that is better.

**Comment by emanjon** [comment]
I updated the PR based on the suggestion from Marco and Göran. I also align the text with the other text in the section. New suggested text is

`Implementations MAY support padding when sending messages. Removal of padding is MANDATORY to support when receiving messages.`

As discussed with kid of type int, I don't think mandatory support when recieving is strictly necesary, but might be a good approach anyway.

**Comment by marco-tiloca-sics** [comment]
tl;dr : Thanks, the commit at https://github.com/lake-wg/edhoc/commit/1d0e09a2ca3f701a46e55337d17c16f8e7f8f6dc looks good :-)

> The optional to use per message is already in the PR I think as "may be used".

Yes and that's fine, but that's when describing the message processing from the point of view of the sender.

> Reading "For the sender, optional to use, on a per-message basis. If the sender is surely never going to pad an outgoing message, no need for it to support padding as a sender at all, of course." It seem like you agree with the suggestion proposed at the interim. We understood it as you proposed changes.

The comment referred to Section 7 "Mandatory-to-Implement Compliance Requirements", for which the new text in PR #190 used to say "OPTIONAL to support". The proposed change was to limit that to the sender side (consistent with the above), but mandatory to the recipient side, to ensure it's able to parse the padding if that was included.

> The "may be used" could be changes to "optional...." if that is better.

I think that "may be used" is also fine in those sections about message processing from the point-of-view of the sender.

**Comment by emanjon** [comment]
Seems like we agree on the high level stuff. Me and Göran spotted that "remove" and "discard" might be the wrong words. The requirement should be on being able to parse. Göran will improve the text. After that we are likely ready to merge

**Comment by gselander** [comment]
I made another commit: 4ee9fc02

MTI section now says:

> Implementations MAY support padding of plaintext when sending messages. Implementations MUST support padding of plaintext when receiving messages, i.e. MUST be able to parse padded messages.

Also some editorials. A difficulty to express what length is hidden since "plaintext" is defined as plain text + padding. This lead to the wording "unpadded plaintext":

> PAD = 1*true is padding that may be used to hide the length of the unpadded plaintext

I kept "discard" in the other parts because I didn't find any good word to substitute for. I didn't like "ignore" or similar as that action seemed to indicate "do nothing". So now it says:

> Decrypt CIPHERTEXT_2, see {{asym-msg2-proc}}, and discard padding, if present.

**Comment by emanjon** [comment]
This has been merged. Closing
